### ISSUE TITLE
Something’s wrong with gas calculation for some old transactions

### ISSUE THREAD

**Issue by Har01d** [ethereum/go-ethereum#18331]
Hey,

We’ve stumbled upon this particular transaction: 0xfb200dc759b00dc69c0578fbce9df31d08c97dc33bc82c84c3b92f3071a4f65b (included in block #68412)

If you call `eth.getTransactionReceipt` on this transaction, you’d get the following:

```
eth.getTransactionReceipt('0xfb200dc759b00dc69c0578fbce9df31d08c97dc33bc82c84c3b92f3071a4f65b');
{
  ...
  gasUsed: 37888,
  ...
```

But when you call `debug.traceTransaction`, you’re getting this:

```
debug.traceTransaction('0xfb200dc759b00dc69c0578fbce9df31d08c97dc33bc82c84c3b92f3071a4f65b');
{
  failed: false,
  gas: 62888,
  ...
```

Note there’s a difference in gas usage which is 25000.

This transaction makes two different calls, you can see them there: https://blockchair.com/ethereum/transaction/0xfb200dc759b00dc69c0578fbce9df31d08c97dc33bc82c84c3b92f3071a4f65b

The second call is to account `0x0000000000000000000000000000000000000004` which we think may be causing this issue.

As we understand, there’s a rule that if you’re sending ethers to a new account, you’re paying additional 25000 gas. But as we also understand EIP-158 (https://github.com/ethereum/EIPs/blob/master/EIPS/eip-158.md) has changed this rule.

We find this discrepancy in 25654 different blocks, the first one is #68412 containing the aforementioned transaction and the last one is #2674779 (a block just before #2675000 when the Spurious Dragon was activated, the transaction is https://blockchair.com/ethereum/transaction/0x92cebf7ce8921bea41d68c455b6d04aeee0a451c53e285e41748b013699cfa9d — note that we show it as a failed transaction, because the tracer returns `out of gas`, while etherscan says it’s ok: https://etherscan.io/tx/0x92cebf7ce8921bea41d68c455b6d04aeee0a451c53e285e41748b013699cfa9d).

So the question is — which value is correct? 37888 or 62888? Why does geth show two different values? Or it’s just a bug in `traceTransaction`?

If it’s 62888 that actually would be bad news meaning that geth also shows wrong account balances, etc. If it’s 37888 it’s bad news for us only since that means we’ll have to repopulate the database :)

CC @linuxion, @wbarobinson.

P.S. If you require more data for research, we’d be glad to help — we @Blockchair have a huge and fast database to look for such things :)

**Comment by stevebreault** [comment]
Interesting that there seems to be an issue being discussed here: https://ethereum.stackexchange.com/questions/40572/how-to-get-internal-transactions-from-call-tracer-results-in-geth?rq=1 which references block 700000 in which this contract also appears frequently: 0x7dE5aBa7DE728950C92C57D08e20D4077161f12F

**Comment by holiman** [comment]
I have a suspiscion that the calltracer does not correctly use the chain rules for historic blocks, we need to look into that. I'll do it when time permits (feel free to ping me on this if nothing happens)

**Comment by Har01d** [comment]
Hey, any update on this? In order to build our explorer further we need the tracer to be fixed, so we’d appreciate if you could give this issue some priority.

For example, we’ve recently launched a service allowing researchers and analysts to download our entire database dumps (see https://gz.blockchair.com), but we can’t add Ethereum because of this bug.

**Comment by holiman** [comment]
Regarding `0x0000000000000000000000000000000000000004` -- that's the `identity` precompile, which was used (early on) by solidity as a way of copying data.

**Comment by holiman** [comment]
Could you please try out the new functionality from https://github.com/ethereum/go-ethereum/pull/17914  ? 
Essentially:
`debug.standardTraceBlockToFile("0xe4cea427f37b6a588ea8d0b37f8ab7cd57cf95a7", {txHash:"0x92cebf7ce8921bea41d68c455b6d04aeee0a451c53e285e41748b013699cfa9d", disableMemory:true})`

That will spit out a json-dump of the transaction, and if you post that trace-file somewhere, I can more easily take a look at it and see what seems to be wrong.

**Comment by linuxion** [comment]
Hello @holiman,

The right command was `debug.standardTraceBlockToFile("0x2e20d1cc2fb48a60ff2749201ae1a6a83424de3b3b7ab48cc48d2f99106e817b", {txHash:"0x92cebf7ce8921bea41d68c455b6d04aeee0a451c53e285e41748b013699cfa9d", disableMemory:true})`.

Traced it and uploaded as [gist](https://gist.github.com/linuxion/12c9e0c4899e106a85918542016d11b7).

**Comment by holiman** [comment]
I think the gist cropped the trace -- look at the last line, it's not complete. Gist does that, in order to create larger files you'll have to do some git tricks and git clone the gist (it's actually a repo) and add it via raw `git` commands instead of the web interface. Alternatively, upload it somewhere else ....

**Comment by holiman** [comment]
The last line of the gist is 
```
{"pc":577,"op":132,"gas":"0x4190","gasCost":"0x3","memory":"0x","memSize":480,"stack":["0x36a8915b","0x139","0x60","0xc0","0x120","0xe4cea427f37b6a588ea8d0b37f8ab7cd57cf95a7","0x139595ce9ea0e4c522f66cccbf4c8369f08f68d31caeea020f939b4bd4506fad","0x60","0xc0","0x120","0x160","0x160","0x180","0x1a0","0x1e0","0x80","0x40","0x40","0x1e0","0x40","0x80","0x0","0x4","0x20"],"depth":1,"refund"
```
-- not even valid `json`. Please recheck

**Comment by Har01d** [comment]
@holiman, that’s exactly the issue — `debug.standardTraceBlockToFile` returns broken json for some reason, it’s not that we’ve uploaded it incorrectly. I guess we can simply try to upload the result of `debug.traceTransaction` (with the `disableMemory:true` option) — would that work for you? Thanks!

**Comment by holiman** [comment]
Yes, that would also work. I'll try to find the cause of why `debug.standardTraceBlockToFile` fails :/

**Comment by holiman** [comment]
I made a fix pr, although that shouldn't really account for more than `4096` bytes lost from the result, but maybe that's all that is missing.

**Comment by holiman** [comment]
I think this issue should be resolved by https://github.com/ethereum/go-ethereum/pull/18503 . Could you please check if that solves your issues?

**Comment by Har01d** [comment]
Yep, it seems like the fix has resolved the issue.

It seems like it has also fixed #16693 and #17590.

**Comment by holiman** [comment]
Great, thanks a lot for confirming @Har01d !
